gh-86608: Improve and restructure tarfile examples (#121771)
python/cpython: Doc/library/tarfile.rst

           print("something else.")
   tar.close()

Writing examples
~~~~~~~~~~~~~~~~

How to create and write an archive to stdout using
:data:`sys.stdout.buffer <sys.stdout>` in the *fileobj* parameter
in :meth:`TarFile.add`::

    import sys
    import tarfile

>     with tarfile.open("sample.tar.gz", "w:gz", fileobj=sys.stdout.buffer) as tar:

Reviewer: streaming mode
```suggestion
    with tarfile.open("sample.tar.gz", "w|gz", fileobj=sys.stdout.buffer) as tar:
```

    tar.add("foo", filter=reset)
    tar.close()

How to create and write an archive to stdout using
:data:`sys.stdout.buffer <sys.stdout>` in the *fileobj* parameter
in :meth:`TarFile.add`::

    import sys
    import tarfile

>     with tarfile.open("sample.tar.gz", "w:gz", fileobj=sys.stdout.buffer) as tar:

Reviewer 2: Does it matter if we use `w:gz` or `w|gz` here?

Reviewer: Streaming mode is probably better to recommend, yes.

A

    tar.add("foo", filter=reset)
    tar.close()

> How to create and write the archive to stdout using 

Reviewer 2: The pre-commit hook is failing because of the trailing space ;)

```suggestion
How to create and write the archive to stdout using
```

Author: Ah, nice catch - I was wondering why it was failing CI

:data:`sys.stdout.buffer <sys.stdout>` in the *fileobj* parameter
in :meth:`TarFile.add`::

>     import tarfile
>     import sys
>     files = ["foo.txt", "bar.txt", "quux.txt"]
>     with tarfile.open("sample.tar.gz", "w|gz", fileobj=sys.stdout.buffer) as tar:
>         for file in files:
>             tar.add(file)

Reviewer: For consistency with the earlier examples.
```suggestion
    import sys
    import tarfile
    with tarfile.open("sample.tar", "w", fileobj=sys.stdout.buffer) as tar:
        for name in ["foo", "bar", "quux"]:
            tar.add(file)
```

Author: Updated

    tar.add("foo", filter=reset)
    tar.close()

How to create an uncompressed tar archive from a list of filenames::

   import tarfile
   tar = tarfile.open("sample.tar", "w")
   for name in ["foo", "bar", "quux"]:
       tar.add(name)
   tar.close()

The same example using the :keyword:`with` statement::

    import tarfile
    with tarfile.open("sample.tar", "w") as tar:
        for name in ["foo", "bar", "quux"]:

>             tar.add(name)

Author: Todo: put these two at the top of the examples page (writing section).
